README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
ROSC
====

ROSC is intended to build and parse messages in the Open Sound Control
(OSC) protocol in R.

Currently, only building messages is implemented. ROSC builds string
representations of OSC Messages that can then be passed to an external
function that encodes OSC for transport.

About OSC
---------

Open Sound Control (OSC) is a transport-independent, message-oriented
communication protocol for communication among computers, devices. OSC
is differentiaed from the related encodings XML, JSON, YMAL, MIME and
MIDI by use of regular expressions and pattern-matching for dispatching,
and by temporal and atomicity semantics. This package partially
implements the OSC 1.1 Encoding Specification
<http://cnmat.berkeley.edu/content/open-sound-control-11-encoding-specification>.

<http://opensoundcontrol.org/>

Many applications, most programming environmnts, and numerous devices
(including sensor/actuator interfaces such as arduino) have OSC client
and/or server implementations.

See:  
<http://opensoundcontrol.org/implementations>  
<https://en.wikipedia.org/wiki/Open_Sound_Control#Applications>

Python, Matlab, and many other computing languages have OSC
implmentations.

R does not (yet), so here is a first attempt to remedy that.

Building OSC Messages
---------------------

### Functions

-   `oscMessage` - Build OSC Message from user supplied data, address
    pattern, and optional type tag string  
-   `oscType` - Helper function to calculate an OSC Type Tag String from
    user supplied data
-   `listFlatten` - Helper function to recursively flatten a list (and
    vector list elements) into a single-level list of elements each of
    length 1.

### Not (yet) supported:

-   OSC Time Tags
-   OSC Blobs
-   OSC Bundles
-   arrays

### Not (likely to be) supported (ever):

-   the full `libo` extension of OSC.

See
<http://www.cnmat.berkeley.edu/sites/default/files/attachments/2015_Dynamic_Message_Oriented_Middleware.pdf>
for a description. This should be a separate package.

Parsing (Dispatching) Messages
------------------------------

Not yet implemented. This will require writing functions to dispatch OSC
Address Patterns and convert OSC arguments to R data types.

Example building OSC Messages with `ROSC`
-----------------------------------------

``` r
library(ROSC)

## Example OSC address patterns
address <- "/audio/1/foo"
address.with.wildcards <- "/{th,s}ing/n[2-4]/red*"

## Example data to pack into OSC messages
data1 <- "bird"
data2 <- 6:8
data3 <- list(list(3:4,"apple"), TRUE, list(list(5.1,"foo"),NULL))

## Example of a manual type string for data3
data3_typestring <- ",iiSidSN"

## Example OSC messages
OSC1 <- oscMessage(address = address, data = data1)
OSC2 <- oscMessage(address = address, data = data2, typecomma=FALSE) # remove comma from typstring
OSC3 <- oscMessage(address = address, data = data3) # now with a nested list of mixed data types
OSC4 <- oscMessage(address = address, data = data3, double = "f") # convert doubles to 32bit floats
OSC5 <- oscMessage(address = address, data = data3, logical = "integer") # convert logical to ints
```

Sending OSC over UDP
--------------------

OSC is transport-independant, but it typically sent over UDP, and less
often over TCP. UDP is chosen for many applications because of the lower
latency (at the expenses of a theoretical possibility of occasional lost
packets), and because of the ability to broadcast without the need for
first establishing a connection, simplifying communication.

OSC over UDP in native R
------------------------

Unlike TCP socket connections, UDP is not natively supported in R. UDP
clients and servers can be written in RCPP, but none that implement OSC
compliant packets have yet been built. This should be be on a to-do list
for interested R developers.

OSC over UDP via Bash command `oscchief`:
-----------------------------------------

### Dependencies:

#### 1. pkg-config <https://www.freedesktop.org/wiki/Software/pkg-config/>

Install with Homebrew: `$ brew install pkg-config` or see
<https://www.freedesktop.org/wiki/Software/pkg-config/>

#### 2. liblo <http://liblo.sourceforge.net/>

Install liblo from source (requires make):

    $ cd liblo-x.xx
    $ ./configure --prefix=/usr/local --enable-static
    $ make
    $ make install

#### 3. osccheif <http://hypebeast.github.io/oscchief/>

compile and install oscchief (requires make):

    $ git clone https://github.com/hypebeast/oscchief.git oscchief
    $ cd oscchief
    $ make
    $ sudo make install

`oscchief` Usage

    oscchief Version 0.2.0
    Copyright (C) 2013 Sebastian Ruml <sebastian.ruml@gmail.com>

    usage: oscchief send HOST PORT OSCADDRESS TYPES ARGUMENTS
           oscchief send FILENAME
           oscchief listen PORT

    positional arguments:
        HOST: IP address of the host where you want to send your OSC message
        PORT: Port number
        OSCADDRESS: OSC address where you want to send your message
        TYPES: OSC type tags. Supported types:
            i - 32 Bit integer
            h - 64 Bit integer
            f - 32 Bit float
            d - 64 Bit double
            c - Char
            s - String
            T - True (no argument required)
            F - False (no argument required)
            N - Nil (no argument required)

    optional arguments:
         -h - Shows this help message

    Examples:
        oscchief send 192.168.0.10 7028 /osc/address ssiii some integers 10 12 786
        oscchief send 192.168.0.10 7028 /osc/address TTiFi 643 98
        oscchief send 192.168.0.10 7028 /osc/address
        oscchief send 192.168.0.10 7028 /osc/address TF
        oscchief listen 7028

### R interface to `oscchief`

``` r

# Wrapper for system("oscchief send")
oscchief.send <- function (host="localhost", port=12345, osc="/") {
  osc <- gsub(",","",osc) # strip type tag comma for compatibility with oscchief
  command <- paste("oscchief send", host, port, osc)
  system(command)
}

# Define host and port
HOST <- "255.255.255.255" # this ip address means "all ip addresses on the local subnet"
LOCALHOST <- "localhost" # equivalent to "127.0.0.1", i.e. your own machine
PORT <- 6789 # recipient must listen on this port number.  Use any number in the thousands.

# Send OSC messages

oscchief.send(host=HOST, port=PORT, osc=OSC1)

oscchief.send(host=HOST, port=PORT, osc=OSC2)

oscchief.send(host=HOST, port=PORT, osc=OSC3)

oscchief.send(host=HOST, port=PORT, osc=OSC4)

oscchief.send(host=HOST, port=PORT, osc=OSC5)
```

Listen to OSC Messages coming from R over UDP
---------------------------------------------

### Listen in a terminal with `osccheif`:

Listen on port 6789

    $ oscchief listen 6789

### Listen in R, with a wrapper for `oscchief`?

``` r
## You might think you could write a wrapper for system("oscchief listen") like this, but...

# DO NOT RUN THIS IN R:
#    (DOES NOT RETURN MESSAGE TO R SESSION)
#    (NO WAY TO GRACEFULLY KILL LISTENER ONCE STARTED)
#
# oscchief.listen <- function (port=12346) {
#   command <- paste("oscchief listen", port)
#   system(command)
# }
# PORT <- 12346
# oscchief.listen(PORT)
#
```

### Listen in Max <https://cycling74.com/products/max>

Note:

Max receives both doubles (int64) and 64-bit ints (int64) as 0.0, even
in Max7 running in 64 bit mode.  
Max receives floats (float32) correctly. TODO: Check code for `udpsend`.
KLUDGE: Send doubles to Max as float32 (OSC Type “f”):

#### R Code (send to Max)

``` r
ADDRESS <- "/audio/1/foo"
DATA <- list(list(3:4,"apple"), TRUE, list(list(5.1,"foo"),NULL))
OSC <- oscMessage(address = address, data = data3, double = "f") # convert doubles to 32bit floats
HOST <- "localhost"
PORT <- 5678
oscchief.send(host=HOST, port=PORT, osc=OSC)
```

#### Max Code (listen)

Copy entire code block into empty Max patcher:

<pre><code>
----------begin_max5_patcher----------
282.3ocoQsraBCCD7ryWgkOmhRLDBzekppJSxpViB1V9QZPH92qe1RendgKq
zNd1clY8kJD4fbALD7i3mvHzkJDJBE.P4dD4DaYXhYhzHB3c4gij5zSVXwFg
UZtvVPM1ySPDtfHbm3hIvF2QaFTwrCuwEu9hFFrIOP6W0Ti62Epq2DpT5pF7
y4I3iws5MvCzaVszYK6tI.dspJTpuuH4FUdiA7Y.2sseW48jX1yJHYYB4S6c
u4l1Ep62+e4t8uyc6M4NNFYhK94OaT3.92OFFoSOTRSVC7WhOBFKWvrbo3FN
zDmeIKSolAsISNJg+ZeTpCs80wVtH0tN1pgYdgeWDgo82Gq+33zoy4x1Mjzn
xQPKb7bj8Jes5CvurQJU
-----------end_max5_patcher-----------
</code></pre>


NEWS.md
# ROSC 0.0.1

* 


R/oscType.R
#' Construct OSC Type Tag String from Input Data
#'
#' Constructs an Open Sound Control (OSC) Type Tag String from the R data types of the input data.
#'
#'     The input is first converted into a one-level list with all elements length 1 using \code{\link{listFlatten}},
#'     then a string is assembled containing one type tag for each element in the input data.
#'     \code{oscType} detects the R object data type and assigns the OSC style type tag specified for that
#'     dataype in the arguments. Currently, only R data types "integer," "double," "character,"
#'     "NULL," and "logical" are supported. R type "logical" always produces type tag "T" or "F".
#'
#'     An OSC Type Tag String is a string of single character type tags preceded by the "," character.  If a function cannot accept
#'     OSC string representations using comma-initial type strings, the comma may be ommitted with \code{typecomma=FALSE}.
#'     Warning: this would be a non-standard representation of OSC.  Any OSC-compliant function expecting a comma-inital type string may
#'     interpret a type string with comma removed as the first data argument and attempt to construct a new type string dynamically
#'     with heuristic methods.
#'
#'     \code{oscType} is agnostic about the interpretation of type tags.  You can use \code{oscType} to convert R datatypes to
#'     tags conforming to OSC 1.0, OSC 1.1 or libo / odot, for example, or even to non-standard tags.
#'
#'     For reference, the OSC 1.0 and related "libo" typ tag specifications are included here:
#'
#'     \strong{OSC 1.0 TYPE TAGS, including extended types}
#'     \url{http://opensoundcontrol.org/spec-1.0}
#'     \tabular{cl}{
#'     \strong{tag} \tab \strong{definition}\cr
#'     i \tab 32-bit signed integer (big-endian 2'S compliment; equivalent of R type "integer")\cr
#'     u \tab 32-bit unsigned integer (big-endian)\cr
#'     h \tab 64 bit signed integer (big-endian two’s complement)\cr
#'     f \tab 32 bit float (big-endian IEEE 754)\cr
#'     d \tab 64 bit double (IEEE 754; equivalent of R type "double")\cr
#'     c \tab an ascii character, sent as 32 bits\cr
#'     r \tab 32 bit RGBA color stored, each channel stored as an 8-bit value: \emph{e.g.:}\cr
#'       \tab \emph{rgba(1.0,0.8,0.0,1.0) = 11111111 11001100 00000000 11111111 = 4291559679}\cr
#'     m \tab 4 byte MIDI message. Bytes from MSB to LSB are: port id, status byte, data1, data2\cr
#'     s \tab OSC-string (sequence of non-null ASCII characters followed by a null, followed by\cr
#'       \tab 0-3 additional null characters to make the total number of bits a multiple of 32.)\cr
#'     S \tab Alternate type represented as an OSC-string\cr
#'       \tab \emph{e.g. for systems that differentiate "symbols" from "strings"}\cr
#'     b \tab OSC-blob (int32 size count n, followed by n 8-bit bytes of arbitrary binary data,\cr
#'       \tab followed by 0-3 zero bytes to make total number of bits a multiple of 32.)\cr
#'     T \tab True (No data argument is passed)\cr
#'     F \tab False (No data argument is passed)\cr
#'     N \tab Nil or Null (No data argument is passed)\cr
#'     I \tab Infinitum (or "Impulse")\cr
#'     [ \tab Indicates the beginning of an array. The tags following are for data in the Array until a close brace tag is reached.\cr
#'     ] \tab Indicates the end of an array.\cr
#'     t \tab OSC-timetag (64-bit big-endian fixed-point NTP timestamp.)\cr
#'       \tab First 32 bits specify the number of seconds since midnight on January 1, 1900.\cr
#'       \tab Last 32 bits specify fractional parts of a second. (Precision is ~200 picoseconds.)\cr
#'       \tab Special case: 63 zero bits followed by one in least significant bit (000...0001)\cr
#'       \tab means "immediately."
#'     }
#'
#'     \strong{libo TYPE TAGS (an extension of OSC 1.0)}
#'     \url{http://www.cnmat.berkeley.edu/sites/default/files/attachments/2015_Dynamic_Message_Oriented_Middleware.pdf}
#'     \tabular{cl}{
#'     \strong{tag} \tab \strong{definition}\cr
#'     c \tab 8-bit signed integer (big-endian  2'S compliment)\cr
#'     C \tab 8-bit unsigned integer (big-endian  2'S compliment)\cr
#'     u \tab 16-bit signed integer (big-endian  2'S compliment)\cr
#'     U \tab 16-bit unsigned integer (big-endian  2'S compliment)\cr
#'     i \tab 32-bit signed integer (big-endian  2'S compliment; equivalent of R type "integer")\cr
#'     I \tab 32-bit unsigned integer (big-endian  2'S compliment)\cr
#'     h \tab 64 bit signed integer (big-endian two’s complement)\cr
#'     H \tab 64 bit unsigned integer (big-endian two’s complement)\cr
#'     d \tab 64 bit double (IEEE 754; equivalent of R type "double")\cr
#'     s \tab OSC-string\cr
#'     S \tab Alternate type represented as an OSC-string\cr
#'     b \tab OSC-blob\cr
#'     B \tab odot bundle\cr
#'     A \tab executable code\cr
#'     T \tab True (No data argument is passed)\cr
#'     F \tab False (No data argument is passed)\cr
#'     N \tab Nil or Null (No data argument is passed)\cr
#'     t \tab OSC-timetag
#'     }
#'
#' @param data list or vector
#' @param integer character. Tag to use for R data type "integer". Default = "i", specifying a 32-bit signed big-endian 2'S compliment integer. (Identical to the R "integer" type.)
#' @param double character. Tag to use for R data type "integer". Default = "d", specifying a 64 bit IEEE 754 double (Identical the R "double" type.)
#' @param character character. Tag to use for R data type "character". Default = "s", specifying a string.
#' @param null character. Tag to use for R data type "character". Default = "s", specifying a string. ="N", typecomma=TRUE
#' @param typecomma logical. If typecomma=TRUE (default), a comma (",") is prepended to the type tag string.
#'
#' @return character. An "OSC Type Tag String." e.g. ",iisdTNf"
#'
#' @examples
#' oscType(7.1)
#' oscType(7:9)
#' oscType(c("foo","gorp"))
#' data <- list(list(3:4,"apple"), TRUE, list(list(5.1,"foo"),NULL))
#' oscType(data)
#' oscType(data, double = "f", typecomma=FALSE)
#'
#' @export
oscType <- function(data=NULL, integer="i", double = "d", character="s", null="N", typecomma=TRUE) {
  if (is.null(data)) {
    return(NULL)
  } else {

    ## flatten data
    data <- listFlatten(data)

    ## extract data types
    t <- sapply(data,typeof)

    ## map input types to OSC data types
    for (i in 1:length(t)) {
      if (is.logical(data[[i]])) {
        if (data[[i]] == TRUE) {
          t[i] <- "T"
        } else {
          t[i] <- "F"
        }
      }
    }

    t <- gsub("integer",integer,t)
    t <- gsub("double",double,t)
    t <- gsub("character",character,t)
    t <- gsub("NULL",null,t)

    ## return OSC Type Tag String
    t <- paste0(t,collapse = "")
    if (typecomma==TRUE) {t <- paste0(",",t)}
    return(t)
  }
}



R/listFlatten.R
#' Flatten list completely
#'
#' Flattens a list recursively, converting vector elements to list elements, producing a one-level list with all elements length 1. Names may not be preserved.
#'
#' @param data list or vector
#'
#' @return single-level list with all list elements of length 1
#'
#' @examples
#' listFlatten(list(list(list(1:5))))
#' listFlatten(7)
#' listFlatten(7:9)
#' listFlatten("foo")
#' listFlatten(c("foo","gorp"))
#' data <- list(list(3:4,"apple"), TRUE, list(list(5.1,"foo"),NULL))
#' listFlatten(data)
#'
#' @export

listFlatten <- function(data) {
  if (!inherits(data, "list") & length(data) < 2) {
    return(list(data))
  } else {
    din <- unlist(c(lapply(data, listFlatten)), recursive = FALSE)
    for (i in 1:length(din)) { if (length(din[[i]]) > 1) din[[i]] <- as.list(din[[i]]) }
    dout <- unlist(c(lapply(din, listFlatten)), recursive = FALSE)
    return(dout)
  }
}

# look at rapply
#
# OLD code depends on rlang:
#
# listFlatten <- function(data) {
#   if (requireNamespace("rlang", quietly = TRUE)) {
#     d <- rlang::squash(as.list(data))
#     for (i in 1:length(d)) {
#       if (!is.null(d[[i]])) {
#         d[[i]] <- as.list(d[[i]])
#       }
#     }
#     d <- rlang::squash(d)
#     return(d)
#   } else {
#     stop('listFlatten requires the rlang package')
#   }
# }



R/oscMessage.R
#' Construct OSC Message from OSC Address Pattern and Data
#'
#' Constructs a string formated as an Open Sound Control (OSC) message, from user supplied OSC Address Pattern, data, and optional OSC Type Tag String.
#'
#'     oscMessage returns a string representation of an OSC message in the format \code{ADDRESS_PATTERN TYPE_TAG_STRING ARGUMENTS}.
#'     (see \url{http://opensoundcontrol.org/spec-1.0} for details.)
#'
#'     It is up to the function accepting this string representation to implement the actual binary encoding of the OSC Message.
#'
#'     Encoded OSC itself is transport-independent, but is often sent over UDP or TCP connections.
#'
#'     \strong{ADDRESS_PATTERN}
#'
#'     The OSC Address Pattern is a forward-slash delimited string represented a heirarchical
#'     user-defined address space (tree), with wildcards for pattern matching.
#'     The minimal OSC Address Pattern "/" represents the root of the tree.
#'     In OSC parlance, each tree node is referred to as a "container"
#'     and the final node is an "OSC method". The address is a means of
#'     dispatching data to a specified method in a heirarchical tree.
#'
#'     \strong{Wildcards:}
#'     \tabular{cl}{
#'     ? \tab matches any single character\cr
#'     * \tab matches any sequence of zero or more characters\cr
#'     \{\} \tab Curly braces containing a comma-separated list of strings (e.g., \{foo,bar\} )\cr
#'          \tab matches any string in the list.\cr
#'     [] \tab Brackets containing a string (e.g., [asdfq]) matches any character in the string.\cr
#'     - \tab (dash) In square brackets, separating two characters (e.g. [a-d] )\cr
#'       \tab matches a range of characters in ASCII collating sequence.\cr
#'     ! \tab In square brackets, preceding a string (e.g. [!xz] ) negates the entire string.\cr
#'       \tab \emph{e.g. 'a[!xz]' matches 'ay' but not 'ax' or 'az')}\cr
#'     }
#'
#'     Parsing of OSC Address Patterns and wildcards and proper dispatching of data is left to the recipient of the OSC Message.
#'
#'     \strong{ARGUMENTS}
#'
#'     Input data is flattened into a one-level list with all elements length 1, converted to a space-delimited argument string
#'     ('ARGUMENTS' in the OSC Message). \code{oscMessage} currently does not support array arguments, OSC-blobs or OSC-bundles.
#'     Nested lists and vectors of length > 1 will all be flattened.
#'
#'     \strong{TYPE_TAG_STRING}
#'
#'     You can supply your own OSC Type Tage String via the \code{typestring} argument. \code{oscMessage} is agnostic about the legality and interpretation
#'     of type tags. You must ensure that the encoding function and end recipeint of the message understand the type tags used. (See below.)
#'
#'     If \code{typestring='auto'}, oscMessage passes the data to \code{\link{oscType}} to generate the string.
#'     The R datatypes of the input data are mapped to OSC type tags as specified by additional arguments to oscMessage.
#'     Currently, only R data types "integer," "double," "character,"
#'     "NULL," and "logical" are supported. See \code{\link{oscType}} for additional details.
#'
#'     A minimal set of OSC type tags for 32 bit integers, 32 bit floats, and strings is standard across implementations.
#'     However, other data types and corresponding tags depend on the OSC version specification and OSC implementation.
#'
#'     For reference, the OSC 1.0 and related "libo" typ tag specifications are included here:
#'
#'     \strong{OSC 1.0 TYPE TAGS, including extended types}
#'     \url{http://opensoundcontrol.org/spec-1.0}
#'     \tabular{cl}{
#'     \strong{tag} \tab \strong{definition}\cr
#'     i \tab 32-bit signed integer (big-endian 2'S compliment; equivalent of R type "integer")\cr
#'     u \tab 32-bit unsigned integer (big-endian)\cr
#'     h \tab 64 bit signed integer (big-endian two’s complement)\cr
#'     f \tab 32 bit float (big-endian IEEE 754)\cr
#'     d \tab 64 bit double (IEEE 754; equivalent of R type "double")\cr
#'     c \tab an ascii character, sent as 32 bits\cr
#'     r \tab 32 bit RGBA color stored, each channel stored as an 8-bit value: \emph{e.g.:}\cr
#'       \tab \emph{rgba(1.0,0.8,0.0,1.0) = 11111111 11001100 00000000 11111111 = 4291559679}\cr
#'     m \tab 4 byte MIDI message. Bytes from MSB to LSB are: port id, status byte, data1, data2\cr
#'     s \tab OSC-string (sequence of non-null ASCII characters followed by a null, followed by\cr
#'       \tab 0-3 additional null characters to make the total number of bits a multiple of 32.)\cr
#'     S \tab Alternate type represented as an OSC-string\cr
#'       \tab \emph{e.g. for systems that differentiate "symbols" from "strings"}\cr
#'     b \tab OSC-blob (int32 size count n, followed by n 8-bit bytes of arbitrary binary data,\cr
#'       \tab followed by 0-3 zero bytes to make total number of bits a multiple of 32.)\cr
#'     T \tab True (No data argument is passed)\cr
#'     F \tab False (No data argument is passed)\cr
#'     N \tab Nil or Null (No data argument is passed)\cr
#'     I \tab Infinitum (or "Impulse")\cr
#'     [ \tab Indicates the beginning of an array. The tags following are for data in the Array until a close brace tag is reached.\cr
#'     ] \tab Indicates the end of an array.\cr
#'     t \tab OSC-timetag (64-bit big-endian fixed-point NTP timestamp.)\cr
#'       \tab First 32 bits specify the number of seconds since midnight on January 1, 1900.\cr
#'       \tab Last 32 bits specify fractional parts of a second. (Precision is ~200 picoseconds.)\cr
#'       \tab Special case: 63 zero bits followed by one in least significant bit (000...0001)\cr
#'       \tab means "immediately."
#'     }
#'
#'     \strong{libo TYPE TAGS (an extension of OSC 1.0)}
#'     \url{http://www.cnmat.berkeley.edu/sites/default/files/attachments/2015_Dynamic_Message_Oriented_Middleware.pdf}
#'     \tabular{cl}{
#'     \strong{tag} \tab \strong{definition}\cr
#'     c \tab 8-bit signed integer (big-endian  2'S compliment)\cr
#'     C \tab 8-bit unsigned integer (big-endian  2'S compliment)\cr
#'     u \tab 16-bit signed integer (big-endian  2'S compliment)\cr
#'     U \tab 16-bit unsigned integer (big-endian  2'S compliment)\cr
#'     i \tab 32-bit signed integer (big-endian  2'S compliment; equivalent of R type "integer")\cr
#'     I \tab 32-bit unsigned integer (big-endian  2'S compliment)\cr
#'     h \tab 64 bit signed integer (big-endian two’s complement)\cr
#'     H \tab 64 bit unsigned integer (big-endian two’s complement)\cr
#'     d \tab 64 bit double (IEEE 754; equivalent of R type "double")\cr
#'     s \tab OSC-string\cr
#'     S \tab Alternate type represented as an OSC-string\cr
#'     b \tab OSC-blob\cr
#'     B \tab odot bundle\cr
#'     A \tab executable code\cr
#'     T \tab True (No data argument is passed)\cr
#'     F \tab False (No data argument is passed)\cr
#'     N \tab Nil or Null (No data argument is passed)\cr
#'     t \tab OSC-timetag
#'     }
#'
#' @param address character. The OSC Address Pattern (see Details).
#' @param data list or vector.
#' @param logical character. R datatype that data of type 'logical' should be converted to prior to processing. Default: \code{logical='logical'}
#' @param typestring character. OSc Type Tag String: a string of non-space characters representing data types, beginning with ",".
#' @param integer character. Tag to use for R data type "integer". Default = "i", specifying a 32-bit signed big-endian 2'S compliment integer. (Identical to the R "integer" type.)
#' @param double character. Tag to use for R data type "integer". Default = "d", specifying a 64 bit IEEE 754 double (Identical the R "double" type.)
#' @param character character. Tag to use for R data type "character". Default = "s", specifying a string.
#' @param null character. Tag to use for R data type "NULL". Default = "N", specifying a "Nil" or null.
#' @param typecomma logocal. If \code{typestring='auto'} and \code{typecomma=TRUE} (the default), a leading comma "," is added to the beginning of the type tag string.
#'
#' @return character. A string representation of an OSC message in the format \code{ADDRESS_PATTERN TYPE_TAG_STRING ARGUMENTS}.
#'
#' @examples
#' address <- "/thing/n1/red"
#' data <- list(list(3:4,"apple"), TRUE, list(list(5.1,"sphere"),NULL))
#' OSC_1 <- oscMessage(address = address, data = data); OSC_1
#'
#' OSC_2 <- oscMessage(address = address, data = data, double = "f"); OSC_2
#'
#' OSC_3 <- oscMessage(address = address, data = data, typecomma=FALSE); OSC_3
#'
#' OSC_4 <- oscMessage(address = address, data = data, logical = "integer"); OSC_4
#'
#' pattern <- "/{th,s}ing/n[2-4]/red*"
#' types <- ",iiSidSN"
#' data <- list(list(3:4,"apple"), TRUE, list(list(5.1,"sphere"),NULL))
#' OSC_5 <- oscMessage(address = pattern, typestring = types, data = data, logical="integer")
#' OSC_5
#'
#' @export
oscMessage <- function(address="/", data=NULL, logical="logical", typestring="auto", integer="i", double = "d", character="s", null="N", typecomma=TRUE) {

  ## flatten data
  d <- data
  if (!is.null(d)) {
    d <- listFlatten(data)
  }

  ## convert logical data
  if (logical == "integer") {
    for (j in 1:length(d)) {
      if (is.logical(d[[j]])) {d[[j]] <- as.integer(d[[j]])}
    }
  }
  if (logical == "double") {
    for (j in 1:length(data)) {
      if (is.logical(d[[j]])) {d[[j]] <- as.double(d[[j]])}
    }
  }

  ## get OSC Type String
  if (typestring=="auto") {
    t <- oscType(
      d, integer=integer, double=double, character=character, null=null,
      typecomma=typecomma)
  } else {
    t <- typestring
  }

  ## contruct OSC argument list from data
  if (!is.null(d)) {
    d <- as.character(d)
    d <- d[!(d %in% c("TRUE","FALSE","NULL"))]
  }

  ## output OSC message
  osc <- paste(address, t, paste(d, collapse = " "), collapse = " ")
  return(osc)
}

### TODO Add time stamp support and full 1.1 spec support
### Accepts data as an atomic vector or list.
### Nested lists will be flattened with listFlatten()

